From d88f22e3b153d7d4b8738af94b9523efceb4b6ba Mon Sep 17 00:00:00 2001
From: Luke Carrier <luke@carrier.family>
Date: Thu, 7 May 2026 11:55:43 +0100
Subject: [PATCH 1/2] fix(workspaces): scale to text width

---
 src/modules/workspaces.rs | 17 ++---------------
 1 file changed, 2 insertions(+), 15 deletions(-)

diff --git a/src/modules/workspaces.rs b/src/modules/workspaces.rs
index 189bcf36..6e159c67 100644
--- a/src/modules/workspaces.rs
+++ b/src/modules/workspaces.rs
@@ -453,27 +453,14 @@ impl Workspaces {
                                         .align_y(alignment::Vertical::Center),
                                 )
                                 .style(theme.workspace_button_style(empty, color))
-                                .padding(if w.id < 0 {
-                                    match w.displayed {
-                                        Displayed::Active => [0, theme.space.md],
-                                        Displayed::Visible => [0, theme.space.sm],
-                                        Displayed::Hidden => [0, theme.space.xs],
-                                    }
-                                } else {
-                                    [0, 0]
-                                })
+                                .padding([0.0, theme.space.xxs])
                                 .on_press(if w.id > 0 {
                                     Message::ChangeWorkspace(w.id)
                                 } else {
                                     Message::ToggleSpecialWorkspace(w.id)
                                 })
-                                .width(match (w.id < 0, &w.displayed) {
-                                    (true, _) => Length::Shrink,
-                                    (_, Displayed::Active) => Length::Fixed(theme.space.xl as f32),
-                                    (_, Displayed::Visible) => Length::Fixed(theme.space.lg as f32),
-                                    (_, Displayed::Hidden) => Length::Fixed(theme.space.md as f32),
-                                })
                                 .height(theme.space.md)
+                                .width(Length::Shrink)
                                 .into(),
                             )
                         } else {
-- 
2.53.0


From 7b4859cbdde3ecbe2c65147ce7fc61736dc5f539 Mon Sep 17 00:00:00 2001
From: Luke Carrier <luke@carrier.family>
Date: Thu, 7 May 2026 12:35:28 +0100
Subject: [PATCH 2/2] feat(workspaces): differentiate between active spaces by
 color

Now that workspaces' buttons are of variable width due to scaling to the
fit the workspace name, we need to visually distinguish between active
and inactive workspaces through some other means. Inverting colors
should let us preserve the existing workspace color options.
---
 src/modules/workspaces.rs |  4 ++--
 src/theme.rs              | 50 ++++++++++-----------------------------
 2 files changed, 14 insertions(+), 40 deletions(-)

diff --git a/src/modules/workspaces.rs b/src/modules/workspaces.rs
index 6e159c67..6540677d 100644
--- a/src/modules/workspaces.rs
+++ b/src/modules/workspaces.rs
@@ -452,8 +452,8 @@ impl Workspaces {
                                         .align_x(alignment::Horizontal::Center)
                                         .align_y(alignment::Vertical::Center),
                                 )
-                                .style(theme.workspace_button_style(empty, color))
-                                .padding([0.0, theme.space.xxs])
+                                .style(theme.workspace_button_style(empty, color, w.displayed == Displayed::Active))
+                                .padding([0.0, theme.space.xxs.into()])
                                 .on_press(if w.id > 0 {
                                     Message::ChangeWorkspace(w.id)
                                 } else {
diff --git a/src/theme.rs b/src/theme.rs
index 2cb6cc15..1058e3f5 100644
--- a/src/theme.rs
+++ b/src/theme.rs
@@ -415,6 +415,7 @@ impl AshellTheme {
         &self,
         is_empty: bool,
         colors: Option<Option<AppearanceColor>>,
+        active: bool,
     ) -> impl Fn(&Theme, Status) -> button::Style {
         move |theme: &Theme, status: Status| {
             let (bg_color, fg_color) = colors.map_or_else(
@@ -443,63 +444,36 @@ impl AshellTheme {
                     )
                 },
             );
-            let mut base = button::Style {
-                background: Some(Background::Color(if is_empty {
-                    theme.extended_palette().background.weak.color
+
+            let base = button::Style {
+                background: Some(Background::Color(if active {
+                    fg_color
                 } else {
                     bg_color
                 })),
                 border: Border {
                     width: if is_empty { 1.0 } else { 0.0 },
-                    color: bg_color,
+                    color: if active { fg_color } else { bg_color },
                     radius: self.radius.lg.into(),
                 },
                 text_color: if is_empty {
                     theme.extended_palette().background.weak.text
+                } else if active {
+                    bg_color
                 } else {
                     fg_color
                 },
                 ..button::Style::default()
             };
+            let mut base = base;
             match status {
                 Status::Active => base,
                 Status::Hovered => {
-                    let (bg_color, fg_color) = colors.map_or_else(
-                        || {
-                            (
-                                theme.extended_palette().background.strong.color,
-                                theme.palette().text,
-                            )
-                        },
-                        |c| {
-                            c.map_or_else(
-                                || {
-                                    (
-                                        theme.extended_palette().primary.strong.color,
-                                        theme.extended_palette().primary.strong.text,
-                                    )
-                                },
-                                |c| {
-                                    let color = palette::Primary::generate(
-                                        c.get_base(),
-                                        theme.palette().background,
-                                        c.get_text().unwrap_or_else(|| theme.palette().text),
-                                    );
-                                    (color.strong.color, color.strong.text)
-                                },
-                            )
-                        },
-                    );
-
-                    base.background = Some(Background::Color(if is_empty {
-                        theme.extended_palette().background.strong.color
-                    } else {
-                        bg_color
-                    }));
+                    base.background = Some(Background::Color(theme.extended_palette().background.strong.color));
                     base.text_color = if is_empty {
-                        theme.extended_palette().background.weak.text
+                        theme.palette().text
                     } else {
-                        fg_color
+                        theme.extended_palette().background.strong.text
                     };
                     base
                 }
-- 
2.53.0

